### PR TITLE
feat: revise uncle rule

### PR DESCRIPTION
**Pull request by zhangsoledad**
1. get rid uncle age limit
2. try include disconnected block as uncle

### CODE REVIEW

**Comment by doitian** [comment]
Add `b:consensus` if this is a breaking change.

File: chain/src/chain.rs [hunk 1]
```diff
@@ -310,6 +310,12 @@ impl<CS: ChainStore + 'static> ChainService<CS> {
                 fork.detached_proposal_id().iter(),
                 &mut txs_verify_cache,
             );
+            if !fork.detached_blocks().is_empty() {
```
Comment (doitian) [inline chain/src/chain.rs]: Seems a redundant check.

File: chain/src/chain.rs [hunk 2]
```diff
@@ -484,31 +490,36 @@ impl<CS: ChainStore + 'static> ChainService<CS> {
...
         let detached_blocks_iter = fork.detached_blocks().iter().rev();
 
-        let unverified_len = fork.attached_blocks.len() - dirty_exts.len();
+        let mut fork_context = ForkContext {
+            attached_blocks: Vec::with_capacity(fork.attached_blocks().len()),
+            detached_blocks: fork.detached_blocks().iter().rev().collect(),
+            provider: self.shared.clone(),
+        };
+
+        let verified_len = fork.attached_blocks.len() - dirty_exts.len();
```
Comment (doitian) [inline chain/src/chain.rs]: 🤣

File: chain/src/chain.rs [hunk 3]
```diff
@@ -484,31 +490,36 @@ impl<CS: ChainStore + 'static> ChainService<CS> {
         let attached_blocks_iter = fork.attached_blocks().iter().rev();
         let detached_blocks_iter = fork.detached_blocks().iter().rev();
 
-        let unverified_len = fork.attached_blocks.len() - dirty_exts.len();
+        let mut fork_context = ForkContext {
+            attached_blocks: Vec::with_capacity(fork.attached_blocks().len()),
+            detached_blocks: fork.detached_blocks().iter().rev().collect(),
```
Comment (doitian) [inline chain/src/chain.rs]: Bad smell, similar structure, same field name, but have a different order.

File: miner/src/block_assembler.rs [hunk 4]
```diff
@@ -135,31 +132,41 @@ impl<CS: ChainStore + 'static> BlockAssembler<CS> {
...
-                        Ok(uncle_block) => {
-                            let hash = uncle_block.header().hash();
-                            self.candidate_uncles.insert(hash.to_owned(), uncle_block);
-                            self.last_uncles_updated_at
-                                .store(unix_time_as_millis(), Ordering::SeqCst);
-                        }
-                        _ => {
-                            log::error!(target: "miner", "new_uncle_receiver closed");
+            .spawn(move || {
+                let mut candidate_uncles = LruCache::new(MAX_CANDIDATE_UNCLES);
```
Comment (doitian) [inline miner/src/block_assembler.rs]: Why move from the field to a local variable? Work around the conflict mutable borrow?

File: verification/src/contextual_block_verifier.rs [hunk 5]
```diff
@@ -20,38 +20,111 @@ use rayon::iter::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIter
...
+        }
+        let detached = fork
+            .detached_blocks
+            .iter()
+            .map(|b| b.header().hash())
+            .collect();
+        let detached_uncles = fork
+            .detached_blocks
+            .iter()
+            .map(|b| b.uncles().iter().map(UncleBlock::hash))
```
Comment (doitian) [inline verification/src/contextual_block_verifier.rs]: can use `flat_map`

**Comment by doitian** [comment]
How can we verify PoW if the uncle is disconnected?

File: miner/src/block_assembler.rs (hunk 4 above)
Comment (zhangsoledad) [reply miner/src/block_assembler.rs]: `Move out `,   `RefCel `l， `Mutex `
I choose  `Move out`

**Comment by zhangsoledad** [comment]
> How can we verify PoW if the uncle is disconnected?
Disconnected block all are PoW verified. 
We process block by order, the orphan block won't appear here.

File: chain/src/chain.rs (hunk 3 above)
Comment (zhangsoledad) [reply chain/src/chain.rs]: I unified the order. haven't idea about name, still .

File: miner/src/block_assembler.rs [hunk 6]
```diff
@@ -429,80 +433,41 @@ impl<CS: ChainStore + 'static> BlockAssembler<CS> {
...
+        loop {
+            if candidate_uncles.is_empty() || uncles.len() == max_uncles_num {
                 break;
             }
-
-            let epoch_number = current_epoch_ext.number();
-
-            // uncle must be same difficulty epoch with candidate
+            let (hash, block) = candidate_uncles
+                .pop_front()
```
Comment (quake) [inline miner/src/block_assembler.rs]: It doesn't insert back the poped uncle block to `candidate_uncles`, this uncle block will be excluded in the next `get_block_template` function invoking.

File: verification/src/contextual_block_verifier.rs [hunk 7]
```diff
@@ -49,11 +49,83 @@ impl<'a, CS: ChainStore> BlockMedianTimeContext for ForkContext<'a, CS> {
...
+
+impl<'a, P: ChainProvider> UncleProvider for UncleVerifierContext<'a, P> {
+    fn double_inclusion(&self, hash: &H256) -> bool {
+        if self.excluded.contains(hash) {
+            return true;
+        }
+
+        // main chain
+        if self.provider.store().get_block_number(hash).is_some() {
+            if self.detached.contains(hash) {
```
Comment (quake) [inline verification/src/contextual_block_verifier.rs]:
```suggestion
            return !self.detached.contains(hash);
```

File: verification/src/contextual_block_verifier.rs [hunk 8]
```diff
@@ -49,11 +49,83 @@ impl<'a, CS: ChainStore> BlockMedianTimeContext for ForkContext<'a, CS> {
...
+                return false;
+            }
+            return true;
+        }
+
+        if self.provider.store().is_uncle(hash) {
+            if self.detached_uncles.contains(hash) {
```
Comment (quake) [inline verification/src/contextual_block_verifier.rs]:
```suggestion
            return !self.detached_uncles.contains(hash);
```

File: verification/src/transaction_verifier.rs [hunk 9]
```diff
@@ -24,13 +25,13 @@ where
     pub fn new(
         rtx: &'a ResolvedTransaction,
         median_time_context: &'a M,
-        tip_number: BlockNumber,
+        block_number: BlockNumber,
         tip_epoch_number: EpochNumber,
```
Comment (keroro520) [inline verification/src/transaction_verifier.rs]: Suggest correcting the variable name `tip_epoch_number` together, and the type should be `EpochNumber`.

**Pull request merged**